contracts/infrastructure/dapp/ParaswapFilter.sol

```diff
@@ -33,55 +36,241 @@ interface IParaswap {
+     * @param _authorised array where each entry is true to add the corresponding DEX to the whitelist, false to remove it
+     */
+    function setAuthorised(address[] calldata _adapters, bool[] calldata _authorised) external onlyOwner {
+        for(uint256 i = 0; i < _adapters.length; i++) {
+            if(adapters[_adapters[i]] != _authorised[i]) {
+                adapters[_adapters[i]] = _authorised[i];
+                if(_authorised[i]) { 
+                    emit AdapterAdded(_adapters[i]); 
+                } else { 
+                    emit AdapterRemoved(_adapters[i]);
+                }
+            }
+        }
     }
 
-    function isMultiSwap(bytes calldata _data) internal pure returns (bool) {
+    function isValid(address _wallet, address /*_spender*/, address _to, bytes calldata _data) external view override returns (bool) {
         // disable ETH transfer
         if (_data.length < 4) {
             return false;
         }
-        return getMethod(_data) == MULTISWAP;
+        bytes4 methodId = getMethod(_data);
+        if(methodId == MULTISWAP) {
+            return isValidMultiSwap(_wallet, _data);
+        } 
+        if(methodId == SIMPLESWAP) {
+            return isValidSimpleSwap(_wallet, _to, _data);
+        }
+        if(methodId == SWAP_ON_UNI) {
+            return isValidUniSwap(_to, _data);
+        }
+        if(methodId == SWAP_ON_UNI_FORK) {
+            return isValidUniForkSwap(_to, _data);
+        }
+        if(methodId == MEGASWAP) {
+            return isValidMegaSwap(_wallet, _data);
+        }
+        return false;
     }
 
-    function hasValidBeneficiary(address _wallet, bytes calldata _data) internal pure returns (bool) {
-        (address beneficiary) = abi.decode(_data[228:], (address)); // skipping 4 + 7*32 = 228 bytes
-        return (beneficiary == address(0) || beneficiary == _wallet);
+    function isValidMultiSwap(address _wallet, bytes calldata _data) internal view returns (bool) {
+        (IParaswap.SellData memory sell) = abi.decode(_data[4:], (IParaswap.SellData));
+        return hasValidBeneficiary(_wallet, sell.beneficiary) &&
+            hasTradableToken(sell.path[sell.path.length - 1].to) && 
+            hasValidExchangeAdapters(sell.path);
     }
 
-    function hasTradableToken(bytes calldata _data) internal view returns (bool) {
-        (, address _destToken) = abi.decode(_data[4:], (address, address));   
+    function isValidSimpleSwap(address _wallet, address _augustus, bytes calldata _data) internal view returns (bool) {
+        (,address toToken,, address[] memory callees,, uint256[] memory startIndexes,, address beneficiary) 
+            = abi.decode(_data[4:], (address, address, uint256[3],address[],bytes,uint256[],uint256[],address));
+        return hasValidBeneficiary(_wallet, beneficiary) &&
+            hasTradableToken(toToken) &&
+            hasAuthorisedCallees(_augustus, callees, startIndexes, _data);
```

Comment:
Do we really need to check the callees given that the calls are executed by AugustusSwapper? In a nutshell what `simpleSwap` does is:
- 1/ transfer the `from` tokens from the wallet to AugustusSwapper
- 2/ execute calls with the callees
- 3/ check that it received the intended amount of `to` tokens
- 4/ transfer the received `to` tokens to the beneficiary

Given the above I have the impression that `hasValidBeneficiary` and `hasTradableToken` combined with check 3/ are sufficient to prevent the wallet from being abused irrespective of what the calls to the callees are. In other words the wallet doesn't care what AugustusSwapper does provided that the above 3 conditions are satisfied.

Reply:
The issue is that check 3/ is not really doable because a compromised owner can pass a `toAmount` value that is as low as they want and trivially satisfy check 3/